pkg/ha/filter.go

```diff
@@ -49,42 +51,125 @@ func (h *Filter) Process(_ *http.Request, wr *prompb.WriteRequest) error {
 
 	minT := model.Time(minTUnix).Time()
 	maxT := model.Time(maxTUnix).Time()
-	allowInsert, acceptedMinT, err := h.service.CheckLease(minT, maxT, clusterName, replicaName)
+	allowInsert, leaseStart, err := h.service.CheckLease(minT, maxT, clusterName, replicaName)
 	if err != nil {
 		return fmt.Errorf("could not check ha lease: %#v", err)
 	}
-	if !allowInsert {
-		log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
-		wr.Timeseries = wr.Timeseries[:0]
+
+	// Short-circuit for no possible backfill data.
+	if !minT.Before(leaseStart) {
+		if !allowInsert {
+			wr.Timeseries = wr.Timeseries[:0]
+			log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
+		}
 		return nil
 	}
 
-	// insert allowed -> parse samples
-	acceptedMinTUnix := int64(model.TimeFromUnixNano(acceptedMinT.UnixNano()))
-	for i := range tts {
-		t := &tts[i]
+	hasBackfill, err := h.filterBackfill(wr, minT, leaseStart, clusterName, replicaName)
+	if err != nil {
+		return fmt.Errorf("could not check backfill ha lease: %#v", err)
+	}
 
-		if minTUnix < acceptedMinTUnix {
-			t.Samples = filterSamples(t.Samples, acceptedMinTUnix)
+	switch {
+	case !hasBackfill && allowInsert:
+		filterOutSampleRange(wr, minTUnix, toPromModelTime(leaseStart))
+	case hasBackfill && !allowInsert:
+		filterOutSampleRange(wr, toPromModelTime(leaseStart), maxTUnix+1) //maxTUnix+1 because filterSamples treats end as exclusive.
+	case !hasBackfill && !allowInsert:
+		wr.Timeseries = wr.Timeseries[:0]
+		log.Debug("msg", "the samples aren't from the leader prom instance. skipping the insert")
+	default:
+		//nothing to do
+	}
+	return nil
+}
+
+func toPromModelTime(t time.Time) int64 {
+	return int64(model.TimeFromUnixNano(t.UnixNano()))
+}
+
+func (h *Filter) filterBackfill(wr *prompb.WriteRequest, minT, maxT time.Time, cluster, replica string) (bool, error) {
+	hasBackfill := false
+	backfillStart := minT
+	for {
+		if !backfillStart.Before(maxT) {
+			break
+		}
+
+		keepRangeStart, keepRangeEnd, err := h.service.GetBackfillLeaseRange(backfillStart, maxT, cluster, replica)
+		if err != nil {
+			return false, fmt.Errorf("could not check backfill ha lease: %#v", err)
+		}
+		// Nothing else to keep.
+		if keepRangeStart.IsZero() || keepRangeEnd.IsZero() {
```

Comment:
Nit: personally don't love testing with IsZero so would prefer the functions return an ok boolean to indicate whether the values are valid, but won't insist on this.

Reply:
I'm torn on this line too. While Go way is to make zero values useful, it does hurt readability in this case.

Will try to improve.